{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' This will only get the data from the weather api, make a few edits, and load it to the local database '''\n",
    "\n",
    "import os\n",
    "import json\n",
    "import time\n",
    "\n",
    "from pyowm import OWM\n",
    "from pyowm.weatherapi25.forecast import Forecast\n",
    "from pyowm.exceptions.api_response_error import NotFoundError\n",
    "from pyowm.exceptions.api_call_error import APICallTimeoutError, APIInvalidSSLCertificateError\n",
    "\n",
    "from pymongo import MongoClient\n",
    "from pymongo.collection import Collection, ReturnDocument\n",
    "from pymongo.database import Database\n",
    "from pymongo.errors import ConnectionFailure, InvalidDocument, DuplicateKeyError, OperationFailure\n",
    "\n",
    "from config import OWM_API_key_loohoo as loohoo_key, OWM_API_key_masta as masta_key\n",
    "from config import port, host, user, password, socket_path\n",
    "\n",
    "\n",
    "def read_list_from_file(filename):\n",
    "    \"\"\" Read the zip codes list from the csv file.\n",
    "        \n",
    "    :param filename: the name of the file\n",
    "    :type filename: sting\n",
    "    \"\"\"\n",
    "    with open(filename, \"r\") as z_list:\n",
    "        return z_list.read().strip().split(',')\n",
    "\n",
    "def get_data_from_weather_api(owm, zipcode=None, coords=None):\n",
    "    ''' Makes api calls for observations and forecasts and handles the API call errors.\n",
    "\n",
    "    :param owm: the OWM API object\n",
    "    :type owm: pyowm.OWM\n",
    "    :param zipcode: the zipcode reference for the API call\n",
    "    :type zipcode: string\n",
    "    :param coords: the latitude and longitude coordinates reference for the API call\n",
    "    :type coords: 2-tuple \n",
    "\n",
    "    returns: the API data\n",
    "    '''\n",
    "    result = None\n",
    "    tries = 1\n",
    "    while result is None and tries < 4:\n",
    "        try:\n",
    "            if coords:\n",
    "                result = owm.three_hours_forecast_at_coords(**coords)\n",
    "            elif zipcode:\n",
    "                result = owm.weather_at_zip_code(zipcode, 'us')\n",
    "        except APIInvalidSSLCertificateError:\n",
    "            loc = zipcode or 'lat: {}, lon: {}'.format(coords['lat'], coords['lon'])\n",
    "            print(f'SSL error with {loc} on attempt {tries} ...trying again')\n",
    "            if coords:\n",
    "                owm_loohoo = OWM(loohoo_key)\n",
    "                owm = owm_loohoo\n",
    "            elif zipcode:\n",
    "                owm_masta = OWM(masta_key)\n",
    "                owm = owm_masta\n",
    "        except APICallTimeoutError:\n",
    "            loc = zipcode or 'lat: {}, lon: {}'.format(coords['lat'], coords['lon'])\n",
    "            print(f'Timeout error with {loc} on attempt {tries}... waiting 1 second then trying again')\n",
    "            time.sleep(1)\n",
    "        tries += 1\n",
    "    if tries == 4:\n",
    "        print('tried 3 times without response; moving to the next step!')\n",
    "        return\n",
    "    return result\n",
    "\n",
    "def get_current_weather(code=None, coords=None):\n",
    "    ''' Get the current weather for the given zipcode or coordinates.\n",
    "\n",
    "    :param code: the zip code to find weather data about\n",
    "    :type code: string\n",
    "    :param coords: the coordinates for the data you want\n",
    "    :type coords: 2-tuple\n",
    "\n",
    "    :return: the raw weather object\n",
    "    :type: json\n",
    "    '''\n",
    "    owm = OWM(loohoo_key)\n",
    "\n",
    "    try:\n",
    "        result = get_data_from_weather_api(owm, zipcode=code)\n",
    "    except APICallTimeoutError:\n",
    "        owm = OWM(loohoo_key)\n",
    "    current = json.loads(result.to_JSON()) # the current weather for the given zipcode\n",
    "    if code:\n",
    "        current['Weather']['zipcode'] = code\n",
    "    current['coordinates'] = current['Location']['coordinates']\n",
    "    current['Weather']['instant'] = 10800*(current['Weather']['reference_time']//10800 + 1)\n",
    "    current['Weather']['time_to_instant'] = current['Weather']['instant'] - current['Weather'].pop('reference_time')\n",
    "    current.pop('Location')\n",
    "    return current\n",
    "\n",
    "def five_day(coords, code=None):\n",
    "    ''' Get each weather forecast for the corrosponding coordinates\n",
    "    \n",
    "    :param coords: the latitude and longitude for which that that weather is being forecasted\n",
    "    :type coords: tuple containing the latitude and logitude for the forecast\n",
    "\n",
    "    :return five_day: the five day, every three hours, forecast for the zip code\n",
    "    :type five_day: dict\n",
    "    '''\n",
    "    owm = OWM(masta_key)\n",
    "\n",
    "    Forecast = get_data_from_weather_api(owm, coords=coords).get_forecast()\n",
    "    forecast = json.loads(Forecast.to_JSON())\n",
    "    if code:\n",
    "        forecast['zipcode'] = code\n",
    "    if coords:\n",
    "        forecast['coordinates'] = coords\n",
    "    forecast.pop('Location')\n",
    "    forecast.pop('interval')\n",
    "    reception_time = forecast['reception_time'] # this is going to be added to the weathers array\n",
    "    for cast in forecast['weathers']:\n",
    "        cast['zipcode'] = forecast['zipcode']\n",
    "        cast['instant'] = cast.pop('reference_time')\n",
    "        cast['time_to_instant'] = cast['instant'] - reception_time\n",
    "    return forecast\n",
    "\n",
    "def dbncol(client, collection, database='test'):\n",
    "    ''' Make a connection to the database and collection given in the arguments.\n",
    "\n",
    "    :param client: a MongoClient instance\n",
    "    :type client: pymongo.MongoClient\n",
    "    :param database: the name of the database to be used. It must be a database name present at the client\n",
    "    :type database: str\n",
    "    :param collection: the database collection to be used.  It must be a collection name present in the database\n",
    "    :type collection: str\n",
    "    \n",
    "    :return col: the collection to be used\n",
    "    :type: pymongo.collection.Collection\n",
    "    '''\n",
    "\n",
    "    db = Database(client, database)\n",
    "    col = Collection(db, collection)\n",
    "    return col\n",
    "\n",
    "def load_og(data, client, database, collection):\n",
    "    # Legacy function...see load_weather() for loading needs\n",
    "    ''' Load data to specified database collection. Also checks for a preexisting document with the same instant and zipcode, and updates\n",
    "    it in the case that there was already one there.\n",
    "\n",
    "    :param data: the dictionary created from the api calls\n",
    "    :type data: dict\n",
    "    :param client: a MongoClient instance\n",
    "    :type client: pymongo.MongoClient\n",
    "    :param database: the database to be used\n",
    "    :type database: str\n",
    "    :param collection: the database collection to be used\n",
    "    :type collection: str\n",
    "    '''\n",
    "    \n",
    "    col = dbncol(client, collection, database)\n",
    "    # create the appropriate filters and update types using the data in the dictionary\n",
    "    if collection == 'instant':\n",
    "        filters = {'zipcode':data['zipcode'], 'instant':data['instant']}\n",
    "        updates = {'$push': {'forecasts': data}} # append the forecast object to the forecasts list\n",
    "        try:\n",
    "            # check to see if there is a document that fits the parameters. If there is, update it, if there isn't, upsert it\n",
    "            updated = col.find_one_and_update(filters, updates, upsert=True, return_document=ReturnDocument.AFTER)\n",
    "#             col.find_one_and_update(filters, updates,  upsert=True)\n",
    "            return\n",
    "        except DuplicateKeyError:\n",
    "            return(f'DuplicateKeyError, could not insert data into {collection}.')\n",
    "    elif collection == 'observed' or collection == 'forecasted':\n",
    "        try:\n",
    "            updated = col.insert_one(data)\n",
    "        except DuplicateKeyError:\n",
    "            return(f'DuplicateKeyError, could not insert data into {collection}.')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_weather(data, client, database, collection):\n",
    "    ''' Load data to specified database collection. This determines the appropriate way to process the load depending on the\n",
    "    collection to which it should be loaded. Data is expected to be a weather-type dictionary. When the collection is \"instants\"\n",
    "    the data is appended the specified object's forecasts array in the instants collection; when the collection is either\n",
    "    \"forecasted\" or \"observed\" the object is insterted uniquely to the specified collection. Also checks for a preexisting\n",
    "    document with the same instant and zipcode, then updates it in the case that there was already one there.\n",
    "\n",
    "    :param data: the dictionary created from the api calls\n",
    "    :type data: dict\n",
    "    :param client: a MongoClient instance\n",
    "    :type client: pymongo.MongoClient\n",
    "    :param database: the database to be used\n",
    "    :type database: str\n",
    "    :param collection: the database collection to be used\n",
    "    :type collection: str\n",
    "    ''' \n",
    "    col = dbncol(client, collection, database=database)\n",
    "    # decide how to handle the loading process depending on where the document will be loaded.\n",
    "    if collection == 'instant' or collection == 'test_instants' or collection == 'instant_temp':\n",
    "        # set the appropriate database collections, filters and update types\n",
    "        if \"Weather\" in data:\n",
    "            updates = {'$set': {'weather': data}} # add the weather to the instant document\n",
    "        else:\n",
    "            updates = {'$push': {'forecasts': data}} # append the forecast object to the forecasts list\n",
    "        try:\n",
    "            filters = {'zipcode':data.pop('zipcode'), 'instant':data.pop('instant')}\n",
    "            col.find_one_and_update(filters, updates,  upsert=True)\n",
    "        except DuplicateKeyError:\n",
    "            return(f'DuplicateKeyError, could not insert data into {collection}.')\n",
    "    elif collection == 'observed' or collection == 'forecasted' or collection == 'obs_temp' or collection == 'cast_temp':\n",
    "        try:\n",
    "            col.insert_one(data)\n",
    "        except DuplicateKeyError:\n",
    "            return(f'DuplicateKeyError, could not insert data into {collection}.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def request_and_load(codes):\n",
    "    ''' Request weather data from the OWM api. Transform and load that data into a database.\n",
    "    \n",
    "    :param codes: a list of zipcodes\n",
    "    :type codes: list of five-digit valid strings of US zip codes\n",
    "    '''\n",
    "    \n",
    "    import make_instants as inst\n",
    "    from make_instants import make_instants\n",
    "    \n",
    "    # Begin a timer for the process and run the request and load process.\n",
    "    start_start = time.time()\n",
    "    print(f'task began at {start_start}')\n",
    "    i, n = 0, 0 # i for counting zipcodes processed and n for counting API calls made; API calls limited to a maximum of 60/minute/apikey.\n",
    "    start_time = time.time()\n",
    "    for code in codes:\n",
    "        try:\n",
    "            current = get_current_weather(code)\n",
    "        except AttributeError:\n",
    "            print(f'got AttributeError while collecting current weather for {code}. Continuing to next code.')\n",
    "            continue\n",
    "        n+=1\n",
    "        coords = current['coordinates']         \n",
    "        try:\n",
    "            forecasts = five_day(coords, code=code)\n",
    "        except AttributeError:\n",
    "            print(f'got AttributeError while collecting forecasts for {code}. Continuing to next code.')\n",
    "            continue\n",
    "        n+=1\n",
    "        load_weather(current, client, 'test', 'obs_temp')\n",
    "        load_weather(forecasts, client, 'test', 'cast_temp')\n",
    "        \n",
    "        # if the api request rate is greater than 60 just keep going. Otherwise check how many requests have been made\n",
    "        # and if it's more than 120 start make_instants.\n",
    "        if n/2 / (time.time()-start_time) <= 1:\n",
    "            i+=1\n",
    "            continue\n",
    "        else:\n",
    "            i+=1\n",
    "            if n>=120:\n",
    "                inst.make_instants(client)\n",
    "                if time.time() - start_time < 60:\n",
    "                    print(f'Waiting {start_time+60 - time.time()} seconds before resuming API calls.')\n",
    "                    time.sleep(start_time - time.time() + 60)\n",
    "                    start_time = time.time()\n",
    "                n = 0\n",
    "\n",
    "    # sort the last of the documents in temp collections\n",
    "    try:\n",
    "        make_instants(client)\n",
    "    except:\n",
    "        print('No more documents to sort into instants')\n",
    "    print(f'task took {time.time() - start_start} seconds and processed {i} zipcodes')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mongodb+srv://chuckvanhoff:Fe7ePrX%215L5Wh6W@cluster0-anhr9.mongodb.net/test?retryWrites=true&w=majority\n",
      "task began at 1587243646.7324429\n",
      "360 360\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 32.41102313995361 seconds before resuming API calls.\n",
      "Timeout error with lat: 36.03, lon: -80.21 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 32.28808808326721 seconds before resuming API calls.\n",
      "Timeout error with lat: 35.91, lon: -79.43 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.25, lon: -79.54 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 36.1, lon: -79.88 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 27.842182874679565 seconds before resuming API calls.\n",
      "Timeout error with lat: 36.12, lon: -78.69 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 36.46, lon: -78.3 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 28.90108609199524 seconds before resuming API calls.\n",
      "Timeout error with lat: 36.43, lon: -78.0 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 27597 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 27605 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 27606 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.76, lon: -78.71 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 14.981015920639038 seconds before resuming API calls.\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 34.42143487930298 seconds before resuming API calls.\n",
      "Timeout error with 27862 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 32.06169390678406 seconds before resuming API calls.\n",
      "Timeout error with 27929 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.14, lon: -75.89 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 36.26, lon: -76.04 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 26.226125955581665 seconds before resuming API calls.\n",
      "Timeout error with lat: 35.35, lon: -80.54 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 29.240548849105835 seconds before resuming API calls.\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 34.25652885437012 seconds before resuming API calls.\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 31.860742330551147 seconds before resuming API calls.\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 30.415770292282104 seconds before resuming API calls.\n",
      "Timeout error with lat: 35.21, lon: -78.1 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.36, lon: -79.11 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.19, lon: -79.47 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 34.99, lon: -79.22 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 34.85, lon: -79.06 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 28379 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 28383 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 6.281578063964844 seconds before resuming API calls.\n",
      "Timeout error with 28445 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 34.77, lon: -78.02 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.12, lon: -77.85 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.46, lon: -77.41 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 14.819917917251587 seconds before resuming API calls.\n",
      "Timeout error with lat: 34.74, lon: -77.38 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 28571 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 28584 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.53, lon: -77.39 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 13.511288166046143 seconds before resuming API calls.\n",
      "Timeout error with lat: 35.65, lon: -81.24 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 23.719595193862915 seconds before resuming API calls.\n",
      "Timeout error with 28701 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.6, lon: -82.29 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.42, lon: -83.44 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.51, lon: -82.84 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 28724 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 28728 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 5.213029861450195 seconds before resuming API calls.\n",
      "Timeout error with 28779 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.35, lon: -83.22 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.64, lon: -82.56 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.6, lon: -82.49 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with 28806 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.62, lon: -82.53 on attempt 1... waiting 1 second then trying again\n",
      "Timeout error with lat: 35.66, lon: -82.49 on attempt 1... waiting 1 second then trying again\n",
      "60 60\n",
      "attempting to move those docs just sorted....I am expecting that this is every document in the cast_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'cast_archive'), that is test.cast_archive\n",
      "attempting to move those docs just sorted....I am assuming that this is every dcoument in the obs_temp collection\n",
      "MOVED docs from Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_temp') to Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'test'), 'obs_archive'), that is test.obs_archive\n",
      "Waiting 6.002815008163452 seconds before resuming API calls.\n",
      "0 0\n",
      "No more documents to sort into instants\n",
      "task took 1080.1368880271912 seconds and processed 1080 zipcodes\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if __name__ == '__main__':\n",
    "    # this try block is to deal with the switching back and forth between computers with different directory names\n",
    "    try:\n",
    "        directory = os.path.join(os.environ['HOME'], 'data', 'forcast-forcast')\n",
    "        filename = os.path.join(directory, 'ETL', 'Extract', 'resources', 'success_zipsNC.csv')\n",
    "        codes = read_list_from_file(filename)\n",
    "    except FileNotFoundError:\n",
    "        directory = os.path.join(os.environ['HOME'], 'data', 'forecast-forecast')\n",
    "        filename = os.path.join(directory, 'ETL', 'Extract', 'resources', 'success_zipsNC.csv')\n",
    "        codes = read_list_from_file(filename)\n",
    "    client = MongoClient(host=host, port=port)\n",
    "    request_and_load(codes)\n",
    "    client.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import make_instants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit ('env': venv)",
   "language": "python",
   "name": "python37364bitenvvenv4a9201e1e72848cf811a9f7702fb1125"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
